commercetools/resource_product_discount.go

```diff
@@ -0,0 +1,408 @@
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: ValidateCurrencyCode,
+									},
+									"cent_amount": {
+										Description: "The amount in cents (the smallest indivisible unit of the currency)",
+										Type:        schema.TypeInt,
+										Required:    true,
+									},
+								},
+							},
+						},
+						"product_id": {
+							Description: "Gift Line Item discount specific field",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						"variant": {
+							Description: "Gift Line Item discount specific field",
+							Type:        schema.TypeInt,
+							Optional:    true,
+						},
+						"supply_channel_id": {
+							Description: "Gift Line Item discount specific field",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						"distribution_channel_id": {
+							Description: "Gift Line Item discount specific field",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+					},
+				},
+			},
+			"predicate": {
+				Description: "A valid [Product Predicate](https://docs.commercetools.com/api/projects/predicates#product-predicates)",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"sort_order": {
+				Description: "The string must contain a number between 0 and 1. All matching product discounts are " +
+					"applied to a product in the order defined by this field. A discount with greater sort order is " +
+					"prioritized higher than a discount with lower sort order. The sort order is unambiguous among all product discounts",
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"is_active": {
+				Description: "When set the product discount is applied to products matching the predicate",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+			},
+			"valid_from": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				DiffSuppressFunc: diffSuppressDateString,
+			},
+			"valid_until": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				DiffSuppressFunc: diffSuppressDateString,
+			},
+			"version": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceProductDiscountCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := getClient(m)
+
+	name := expandLocalizedString(d.Get("name"))
+	description := expandLocalizedString(d.Get("description"))
+
+	value, err := expandProductDiscountValue(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	draft := platform.ProductDiscountDraft{
+		Name:        name,
+		Description: &description,
+		Value:       &value,
+		Predicate:   d.Get("predicate").(string),
+		SortOrder:   d.Get("sort_order").(string),
+		IsActive:    d.Get("is_active").(bool),
+	}
+
+	key := stringRef(d.Get("key"))
+	if *key != "" {
+		draft.Key = key
+	}
```

Comment:
I've added in utils a `nilIEmpty()` function. so you can write `key := nilIfEmpty(stringRef(d.Get("key")))`